[Merged by Bors] - feat(Analysis/InnerProductSpace/Orthogonal): add duplicates for `ClosedSubmodule`

yoh-tanimoto opened the pull request:
add duplicates of definitions and API's for `ClosedSubmodule` in an `InnerProductSpace`.

motivation: when one considers a Hilbert space (`CompleteSpace`), it is more natural to consider `ClosedSubmodule`s. They form a complete lattice and satisfy `sup_orthogonal` #29243.

- [x] depends on: #29230 for `Lattice` `CompleteLattice`

yoh-tanimoto committed:
Update ClosedSubmodule.lean

yoh-tanimoto committed:
Merge branch 'yoh-tanimoto-closedsubmodule' into yoh-tanimoto-closedorthogonal

yoh-tanimoto committed:
Update Orthogonal.lean

yoh-tanimoto committed:
Update Orthogonal.lean

yoh-tanimoto committed:
Merge branch 'master' into yoh-tanimoto-closedorthogonal

yoh-tanimoto committed:
Update Orthogonal.lean

yoh-tanimoto committed:
Merge branch 'master' into yoh-tanimoto-closedorthogonal

sgouezel reviewed (commented):
I have to say I am not completely convinced that all the duplication is useful, as many results are just applications of the Module result. Maybe have a lemma saying that the (orthogonal of K) seen as a Module is the same as the orthogonal of (K seen as a module) to enable rewriting, and don't keep the lemmas that are defeq to their Module versions?

yoh-tanimoto committed:
Merge branch 'master' into yoh-tanimoto-closedorthogonal

yoh-tanimoto committed:
adapt to merge master

yoh-tanimoto committed:
Update ClosedSubmodule.lean

yoh-tanimoto committed:
change def, remove some duplicates

yoh-tanimoto commented:
I tried to eliminate as much duplicates as I could. However, even though I simplified the definition of `ClosedSubmodule.orthogonal`, I don't still have `K.toSubmodule = K'.toSubmodule ↔ K = K'` by `rfl`, but it is rather a consequence of an `ext` lemma. Consequently, I could not reduce many lemmas do to the corresponding lemmas in `Submodule`, but are a combination of `simp` and `exact`. I would be happy to know if there is a better way.

I added some lemmas for rewriting.

yoh-tanimoto committed:
simp direction

yoh-tanimoto committed:
follow lint

yoh-tanimoto committed:
correct simp

yoh-tanimoto committed:
Merge branch 'master' into yoh-tanimoto-closedorthogonal

sgouezel commented on Mathlib/Analysis/InnerProductSpace/Orthogonal.lean:
```diff
@@ -378,3 +384,147 @@ theorem OrthogonalFamily.isOrtho {ι} {V : ι → Submodule 𝕜 E}
+
+/-- When a vector is in `Kᗮ`. -/
+@[simp]
+theorem mem_orthogonal (v : E) : v ∈ Kᗮ ↔ ∀ u ∈ K, ⟪u, v⟫ = 0 := Iff.of_eq rfl
```
```suggestion
theorem mem_orthogonal (v : E) : v ∈ Kᗮ ↔ ∀ u ∈ K, ⟪u, v⟫ = 0 := Iff.rfl
```

yoh-tanimoto committed:
Update Mathlib/Analysis/InnerProductSpace/Orthogonal.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

yoh-tanimoto committed:
Update Mathlib/Analysis/InnerProductSpace/Orthogonal.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

yoh-tanimoto commented on Mathlib/Analysis/InnerProductSpace/Orthogonal.lean:
```diff
@@ -378,3 +384,147 @@ theorem OrthogonalFamily.isOrtho {ι} {V : ι → Submodule 𝕜 E}
+  simpa using mem_orthogonal_iff _ _
+
+/-- In a complete space, the orthogonal complement of any submodule `K` is complete. -/
+instance [CompleteSpace E] : CompleteSpace Kᗮ :=
```
and turned the theorem `IsClosed_orthogonal` in `Submodule` into an instance

yoh-tanimoto committed:
Update Mathlib/Analysis/InnerProductSpace/Orthogonal.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

yoh-tanimoto committed:
Update Orthogonal.lean

yoh-tanimoto committed:
Merge remote-tracking branch 'upstream/master' into yoh-tanimoto-closedorthogonal

yoh-tanimoto committed:
restore instance and postpone the removal to #33693

sgouezel commented on Mathlib/Analysis/InnerProductSpace/Orthogonal.lean:
```diff
@@ -378,3 +384,138 @@ theorem OrthogonalFamily.isOrtho {ι} {V : ι → Submodule 𝕜 E}
+
+theorem sub_mem_orthogonal_of_inner_right {x y : E} (h : ∀ v : K, ⟪(v : E), x⟫ = ⟪(v : E), y⟫) :
+    x - y ∈ Kᗮ := by
+  exact Submodule.sub_mem_orthogonal_of_inner_right h
```
```suggestion
    x - y ∈ Kᗮ :=
  Submodule.sub_mem_orthogonal_of_inner_right h
```
Most of the time, `by exact` is useless.

yoh-tanimoto replied:
done, with another unnecessary `by exact`

yoh-tanimoto commented on Mathlib/Analysis/Convex/Cone/Basic.lean:
```diff
@@ -80,8 +80,6 @@ instance : SetLike (ProperCone R E) E where
 
 @[ext] lemma ext (h : ∀ x, x ∈ C₁ ↔ x ∈ C₂) : C₁ = C₂ := SetLike.ext h
 
-@[simp] lemma mem_toPointedCone : x ∈ C.toPointedCone ↔ x ∈ C := .rfl
```
I see, removed attrib

yoh-tanimoto committed:
restore lemma, remove `by exact`

yoh-tanimoto committed:
revert instance to theorem

sgouezel commented on Mathlib/Analysis/InnerProductSpace/Orthogonal.lean:
```diff
@@ -378,3 +384,134 @@ theorem OrthogonalFamily.isOrtho {ι} {V : ι → Submodule 𝕜 E}
+/-- When a vector is in `Kᗮ`, with the inner product the
+other way round. -/
+theorem mem_orthogonal' (v : E) : v ∈ Kᗮ ↔ ∀ u ∈ K, ⟪v, u⟫ = 0 :=
+  Submodule.mem_orthogonal' (K.toSubmodule) v
```
```suggestion
  Submodule.mem_orthogonal' K.toSubmodule v
```

yoh-tanimoto committed:
Update Mathlib/Analysis/InnerProductSpace/Orthogonal.lean

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

yoh-tanimoto committed:
Apply suggestions from code review

Co-authored-by: Sebastien Gouezel <sebastien.gouezel@univ-rennes1.fr>

yoh-tanimoto committed:
restore the previous version

yoh-tanimoto committed:
remove `simp` attrib

sgouezel commented on Mathlib/Analysis/InnerProductSpace/Orthogonal.lean:
```diff
@@ -378,3 +384,134 @@ theorem OrthogonalFamily.isOrtho {ι} {V : ι → Submodule 𝕜 E}
+/-- `orthogonal.orthogonal` preserves the `≤` ordering of two
+subspaces. -/
+theorem orthogonal_orthogonal_monotone {K₁ K₂ : ClosedSubmodule 𝕜 E} (h : K₁ ≤ K₂) : K₁ᗮᗮ ≤ K₂ᗮᗮ :=
+  orthogonal_le (orthogonal_le h)
```
Ah, you're right, I was wrongly thinking of a complete space. So it's perfectly fine to keep the lemma, sorry for the noise!

sgouezel commented:
bors r+
Thanks!
